### PR TITLE
Finished Paper Form Component

### PR DESCRIPTION
**Pull request by shoxter**
Usage like this:

```
{{#paper-form parentSubmit=(action 'useless') as |form|}}

         {{form.input value=foo onChange=(action (mut foo)) label="Foo" required=true}}
         {{form.input value=bar onChange=(action (mut bar)) label="Bar"}}

         {{#paper-button primary=true raised=true onClick=form.submit disabled=form.isInvalid}}Button{{/paper-button}}

{{/paper-form}}
```

Different Options: 

- **form.input:** Paper inputs inside of the form.
- **parentSubmit:**  The action in the parent component/controller to execute (clear fields, create/save records, etc) with regards to the form prior to the form component resetting the isTouched property on all of the inputs.
- **form.submit:** This should be the onClick method for the intended submit button for the form.
- **form.isInvalid:** This is the property that the form changes as the form's inputs change their validity state. When the inputs are invalid, this will be true and should disabled the form's intended submit button.

The form looks at all the form.inputs with input constraints for invalidity.

Please post any questions. We can create a page on the dummy app to cover the documentation for this component once we cover any questions that should be included.

### CODE REVIEW

**Comment by miguelcobain** [comment]
Should I consider this PR or https://github.com/miguelcobain/ember-paper/pull/406 ?

**Comment by shoxter** [comment]
@miguelcobain Both of them.

Multiple input validation in a form setting wasn't possible without the fix to inputs in #406.

I first had to fix the inputs to not send separate isInvalid calls for both 'NULL' and 'TRUE' values.

It was effectively double registering the invalid state of the input. #406 fixes that and this is the form component.

**Comment by DanChadwick** [comment]
Maybe we should have `isValad` so null is better aligned with false.

File: addon/components/paper-form.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ export default Ember.Component.extend({
 	tagName: '',
 	classNames: [],
 	attributeBindings: ['style'],
-	style: "width:100%",
+	style: 'width:100%',
 	numberOfInvalids: 0,
 	isValid: Ember.computed('numberOfInvalids', function() {
 		return this.get('numberOfInvalids') === 0 ? true : false;
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: `return condition ? true : false;` is equivalent to `return condition;`

File: addon/components/paper-form.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+	tagName: '',
+	classNames: [],
+	attributeBindings: ['style'],
+	style: 'width:100%',
+	numberOfInvalids: 0,
+	isValid: Ember.computed('numberOfInvalids', function() {
+		return this.get('numberOfInvalids') === 0 ? true : false;
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: `return condition ? true : false;` is equivalent to `return condition;`

File: addon/components/paper-form.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: destructure all the things

File: addon/components/paper-form-container.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+	tagName: '',
+	attributeBindings: ['style'],
+	style: 'width:100%',
+	customValidations: [],
+	hasValidation: Ember.computed('min', 'max', 'errorMessages', 'maxlength', 'required', 'customValidations', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form-container.js]: destructure all the things

File: addon/components/paper-form-container.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+		return this.get('min') || this.get('max') || this.get('errorMessages') || this.get('maxlength') || this.get('required') || this.get('customValidations');
+	}),
+	didUpdateAttrs: function(params) {
```
Comment (miguelcobain) [inline addon/components/paper-form-container.js]: use shorthand syntax `didUpdateAttrs(params) {`

File: addon/components/paper-form-container.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+		if (params.oldAttrs.isTouched === params.newAttrs.isTouched) return;
+		this.set('touched', this.get('isTouched'));
+	},
+	actions: {
+		callOnInvalid: function(status) {
```
Comment (miguelcobain) [inline addon/components/paper-form-container.js]: shorthand syntax here as well

File: addon/components/paper-form.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+export default Component.extend({
+  tagName: '',
+  classNames: [],
+  attributeBindings: ['style'],
+  style: 'width:100%',
+  numberOfInvalids: 0,
+  isValid: Ember.computed('numberOfInvalids', function() {
+    return this.get('numberOfInvalids') === 0;
+  }),
+  isInvalid: Ember.computed('isValid', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: destructure `computed`

File: addon/components/paper-form.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    return !this.get('isValid');
+  }),
+  sendToParent: Ember.on('init', Ember.observer('isValid', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: destructure `on`

Comment (miguelcobain) [inline addon/components/paper-form.js]: destructure `observer`

File: addon/components/paper-form-container.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import Ember from 'ember';
+const { Component } = Ember;
+
+export default Component.extend({
+  tagName: '',
+  attributeBindings: ['style'],
+  style: 'width:100%',
+  onChange: null,
+  customValidations: [],
+  hasValidation: Ember.computed('min', 'max', 'errorMessages', 'maxlength', 'required', 'customValidations', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form-container.js]: destructure `computed`

File: addon/components/paper-form.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import Ember from 'ember';
+const { Component } = Ember;
+
+export default Component.extend({
+  tagName: '',
+  classNames: [],
+  attributeBindings: ['style'],
+  style: 'width:100%',
+  numberOfInvalids: 0,
+  isValid: Ember.computed('numberOfInvalids', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: destructure `computed`

**Comment by miguelcobain** [comment]
can't you work around the `touchedTrigger` by using `_isTouched` instead of `isTouched` on the templates?

**Comment by shoxter** [comment]
@miguelcobain No because the value will never change in the parent component. The form is never going to set isTouched = true. So oldAttrs will always = newAttrs and will never execute. Unless I call a this.set('isTouched', true) then this.set('isTouched', false) which is more hacky than this solution.

**Comment by miguelcobain** [comment]
Can you edit this twiddle to reproduce that problem?
https://ember-twiddle.com/5d15b9f7ad4092d7bfabc718751c2137?openFiles=components.my-component.js%2Ctemplates.components.my-component.hbs

**Comment by shoxter** [comment]
@miguelcobain https://ember-twiddle.com/5d15b9f7ad4092d7bfabc718751c2137?openFiles=controllers.application.js%2Ctemplates.components.my-component.hbs

**Comment by miguelcobain** [comment]
How do I reproduce?

**Comment by shoxter** [comment]
@miguelcobain It didn't save for whatever reason. 

This is the right one: https://ember-twiddle.com/d7bc618f48a24a89e1bd9411c0d49050/91f4de4c453bcbe8650134d9d72c026ead80cf45?openFiles=controllers.application.js%2C

1.) Toggle controller.
2.) Toggle component.
3.) Toggle controller again.

File: addon/components/paper-form.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import ParentMixin from 'ember-paper/mixins/parent-mixin';
+
+const { Component, computed, on, observer } = Ember;
+
+export default Component.extend(ParentMixin, {
+  tagName: '',
+  isValid: computed('childComponents.@each.isInvalid', function() {
+    return !this.get('childComponents').isAny('isInvalid');
+  }),
+  isInvalid: computed('isValid', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: We can use `computed.not` here

File: addon/mixins/child-mixin.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import Ember from 'ember';
+const { Mixin } = Ember;
+
+export default Mixin.create({
+  asdfgh: 'lel',
```
Comment (miguelcobain) [inline addon/mixins/child-mixin.js]: Maybe we should remove this. :)

Comment (shoxter) [inline addon/mixins/child-mixin.js]: Fixed.

File: addon/components/paper-form.js (hunk 11 above)
Comment (shoxter) [inline addon/components/paper-form.js]: Done.

File: addon/mixins/child-mixin.js (hunk 12 above)
Comment (DanChadwick) [inline addon/mixins/child-mixin.js]: OH NO! My code depended upon `asdfgh`. :stuck_out_tongue_winking_eye:

Comment (shoxter) [inline addon/mixins/child-mixin.js]: Lel!

File: addon/components/paper-form.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+import ParentMixin from 'ember-paper/mixins/parent-mixin';
+
+const { Component, computed, on, observer } = Ember;
+
+export default Component.extend(ParentMixin, {
+  tagName: '',
+  isValid: computed.not('childComponents.@each.isInvalid', function() {
+    return this.get('childComponents').isAny('isInvalid');
+  }),
+  isInvalid: computed.not('isValid', function() {
```
Comment (miguelcobain) [inline addon/components/paper-form.js]: Please read docs on computed.not. it doesn't take a callback

Comment (shoxter) [inline addon/components/paper-form.js]: Sorry about that. I was in a rush out the door. I'll fix it when I get back.

**Comment by dustinfarris** [comment]
If I want to use a custom paper-input component, is there an easy way to plug that in here?

**Comment by shoxter** [comment]
`
{{#paper-form as |form|}}
  {{custom-paper-input parentComponent=form}}
{{/paper-form}}
`

**Comment by DanChadwick** [comment]
@dustinfarris: I agree. Any component yielded should be created from an attribute. Contextual components tightly couple. :-1:

**Comment by shoxter** [comment]
@DanChadwick The paper-input is an attribute of the contextual component because it's the most common input yielded in the form.

It's impossible to take into account all possibilities in the contextual component (which I think is what you're saying), but you can make any paper-input based component a part of the form by adding `parentComponent=form` inside of the form component.

**Comment by dustinfarris** [comment]
@shoxter thanks for that example.  what if i'm using an intermediary component that proxies to paper-input?  e.g. my usage here https://github.com/DockYard/ember-changeset/issues/59#issuecomment-231567141

do i pass parentComponent down the chain?  or just to the first component?  is this officially supported?

**Comment by shoxter** [comment]
@dustinfarris No problem.

The reason `parentComponent=form` works is because the form looks at all children components (components that you've specified that their parent is the form) for isInvalid. 

`isInvalid` is a property of paper-input, so unless your proxy component has a property that aliases the isInvalid property of the paper-input in your template, you'll need to have it on the paper-input.

Make sense?

**Comment by dustinfarris** [comment]
@shoxter i think so, i'm just wondering how this will work with an intermediary component (i guess i'll find out soon enough!)

e.g.

#### my-route.hbs

```hbs
{{#paper-form as |form|}}
  {{x-paper-input parentComponent=form}}
{{/paper-form}}
```

#### x-paper-input.hbs

```hbs
{{paper-input parentComponent=parentComponent}}
```

**Comment by miguelcobain** [comment]
@dustinfarris x-paper-input extends paper-input?

**Comment by dustinfarris** [comment]
@miguelcobain no, it is a wrapper

**Comment by dustinfarris** [comment]
@miguelcobain see Slack for extended conversation on this

**Comment by miguelcobain** [comment]
@dustinfarris should be fixed with https://github.com/miguelcobain/ember-paper/commit/35d457fac0f18dd562db25e14d4d7d85e1741620

**Pull request merged**